pinot-controller/src/main/java/com/linkedin/pinot/controller/util/TableRetentionValidator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.helix.PropertyPathConfig;
+import org.apache.helix.PropertyType;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The <code>TableRetentionValidator</code> class validates the retention policy in table config, and the start/end
+ * timestamp in segment metadata.
+ * <p>Will validate the followings:
+ * <ul>
+ *   <li>
+ *     Table Config
+ *     <ul>
+ *       <li>"segmentsConfig" is set.</li>
+ *       <li>"segmentPushType" is set to APPEND or REFRESH.</li>
+ *       <li>Retention setting is valid for APPEND push type.</li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     Segment Metadata
+ *     <ul>
+ *       <li>For APPEND push type, offline segment start/end time and time unit is valid.</li>
+ *     </ul>
+ *   </li>
+ * </ul>
+ */
+public class TableRetentionValidator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TableRetentionValidator.class);
+
+  private static final String TABLE_CONFIG_PATH = "/CONFIGS/TABLE/";
+  private static final String SEGMENTS_PATH = "/SEGMENTS/";
+
+  private final String _clusterName;
+  private final ZKHelixAdmin _helixAdmin;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+  public TableRetentionValidator(@Nonnull String zkAddress, @Nonnull String clusterName) {
+    _clusterName = clusterName;
+    _helixAdmin = new ZKHelixAdmin(zkAddress);
+    _propertyStore = new ZkHelixPropertyStore<>(zkAddress, new ZNRecordSerializer(),
+        PropertyPathConfig.getPath(PropertyType.PROPERTYSTORE, clusterName));
+  }
+
+  public void run()
+      throws Exception {
+    // Get all resources in cluster
+    List<String> resourcesInCluster = _helixAdmin.getResourcesInCluster(_clusterName);
+
+    for (String tableName : resourcesInCluster) {
+      // Skip non-table resources
+      if (!tableName.endsWith("_OFFLINE") && !tableName.endsWith("_REALTIME")) {
+        continue;
+      }
+      LOGGER.info("Table: {}", tableName);
+
+      // Get the retention config
+      SegmentsValidationAndRetentionConfig retentionConfig = getTableConfig(tableName).getValidationConfig();
+      if (retentionConfig == null) {
+        LOGGER.error("ERROR: \"segmentsConfig\" field is missing in table config");
+        continue;
+      }
+      String segmentPushType = retentionConfig.getSegmentPushType();
+      if (segmentPushType == null) {
+        LOGGER.error("ERROR: null push type");
+        continue;
+      } else if (segmentPushType.equalsIgnoreCase("REFRESH")) {
+        LOGGER.info("Push type: REFRESH");
+        continue;
+      } else if (segmentPushType.equalsIgnoreCase("APPEND")) {
+        LOGGER.info("Push type: APPEND");
+      } else {
+        LOGGER.error("ERROR: Invalid push type: {}", segmentPushType);
+        continue;
```

Comment:
For this and some errors below that validate retention time/unit, it's better to proceed further to get error information on segment metadata. Otherwise we'll have to run this multiple times after fixing each problem